src/future.rs

```diff
@@ -75,26 +76,57 @@ enum RetryFuturePoll<A> where A: Action {
 
 /// Future that drives multiple attempts at an action via a retry strategy.
 pub struct Retry<I, A> where I: Iterator<Item=Duration>, A: Action {
+    retry_if: RetryIf<I, A, fn(&A::Error) -> bool>
+}
+
+impl<I, A> Retry<I, A> where I: Iterator<Item=Duration>, A: Action {
+    pub fn spawn<T: IntoIterator<IntoIter=I, Item=Duration>>(handle: Handle, strategy: T, action: A) -> Retry<I, A> {
+        fn retry_always<A: Action>(_error: &A::Error) -> bool { true };
+        Retry {
+            retry_if: RetryIf::with_condition(handle, strategy, action, retry_always::<A>)
+        }
+    }
+}
+
+impl<I, A> Future for Retry<I, A> where I: Iterator<Item=Duration>, A: Action {
+    type Item = A::Item;
+    type Error = Error<A::Error>;
+
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        self.retry_if.poll()
+    }
+}
+
+/// Future that drives multiple attempts at an action via a retry strategy. Retries are only attempted if
+/// the `Error` returned by the future satisfies a given condition.
+pub struct RetryIf<I, A, C> where I: Iterator<Item=Duration>, A: Action, C: Condition<A::Error> {
     strategy: I,
     state: RetryState<A>,
     action: A,
-    handle: Handle
+    handle: Handle,
+    condition: C
 }
 
-impl<I, A> Retry<I, A> where I: Iterator<Item=Duration>, A: Action {
-    pub fn spawn<T: IntoIterator<IntoIter=I, Item=Duration>>(handle: Handle, strategy: T, mut action: A) -> Retry<I, A> {
-        Retry {
+impl<I, A, C> RetryIf<I, A, C> where I: Iterator<Item=Duration>, A: Action, C: Condition<A::Error> {
+    pub fn with_condition<T: IntoIterator<IntoIter=I, Item=Duration>>(
```

Comment:
As you mentioned earlier, for consistency's sake this could be just `spawn` now that it lives on another struct.

I've also thought about deprecating the constructors, and just exposing a `retry` function, but that's for another day...